documentation/projects/proposals/dark_mode/20240325-implementation_plan_dark_mode.md

```diff
@@ -0,0 +1,263 @@
+milestones like when a feature flag could be made available in a particular environment.
+-->
+
+The following plan requires approved designs and semantic color names. Each task
+is a discrete issue and pull request. The top-level "Work Streams" can be
+completed in parallel.
+
+1. **Work Stream A**: Implement the new color palette.
+
+   1. Create a `FORCE_DARK_MODE` feature flag which is "off" by default and
+      "switchable" in our staging environment. Add a `dark-mode` class to the
+      root HTML tag when this flag is enabled, and a `light-mode` class which is
+      set by default. This will not result in any visual changes.
+   2. The following steps can take place in parallel:
+
+      1. Rename all colors in the Tailwind configuration and the frontend
+         components to use new, semantic names (specific names TBD). This will
+         not result in any visual changes. **This is likely to be the largest PR
+         to review as it is a global find/replace across the entire frontend.**
+
+         1. Replace the "hardcoded" color values in the Tailwind configuration
+            file with css variables defined in the "base" layer of the
+            `tailwind.css` file. This will not result in any visual changes.
+
+      2. Add the dark mode colors as CSS variable definitions, nested under the
+         `.dark-mode` CSS class, in the "base" layer of the `tailwind.css` file.
+         This will not result in any visual changes, _except when
+         `FORCE_DARK_MODE` is enabled._
+
+      3. Visual Regression tests. Update
+         `frontend/test/playwright/utils/breakpoints.ts` so that each breakpoint
+         produces and expects a dark mode screenshot to pass as well as the
+         existing light mode screenshot. **This will also be a significant diff,
+         as at the time of writing it will create 293 new screenshots to review.
+         This is also the point of the process where @fcoveram and
+         @wordpress/openverse-frontend should review the full dark mode
+         appearance for correctness and sufficient color contrast (see the
+         ["Accessibility"](#accessibility) section for more details.**
+
+2. **Work Stream B**: Toggling dark mode
+
+   1. Create a `DARK_MODE_UI_TOGGLE` feature flag which is off by default and
+      switchable in staging.
+   2. Implement logic for calculating a current "color mode" with the following
+      state:
+
+      ```ts
+      interface ColorMode {
+        preference: "dark" | "light" | "system" // Defaults to "light"
+        systemValue: "light" | "dark" // Readonly representation of the system value
+      }
+      ```
+
+      The color mode should be stored in a cookie so that a previously-selected
+      user choice can be used when rendering via SSR and prevent a visual flash
+      of light mode styles for users who have selected dark mode. The system
+      value is read by matching the `'(prefers-color-scheme: dark)'` media
+      query.
```

Comment:
How will our frontend caching affect users with this cookie set? Do we need to consider a `Vary` header on the response for these users?

Reply:
I need to think about this a bit more but it might be wise for us to disable frontend caching on the actual site pages, OR store the color theme preference in a cookie _and_ in localstorage simultaneously :thinking:

Reply:
A cookie is the only option workable for SSR, I think. SSR has no access to local storage, we really should only rely on local storage for state that matters only after users are actively interacting with the page (recent searches, for example) or when we absolutely need to avoid GDPR problems.

Reply:
For this and https://github.com/WordPress/openverse/pull/3963#discussion_r1575399744, it would be helpful to map out the flow and end states for the different scenarios, maybe as Mermaid charts to help with visualisation. I just want to make sure we've covered the cases going in. I believe they are:

- SSR
  - First time visitor
    - prefers-color-scheme: dark -> dark render
    - prefers-color-scheme: light (or unset) -> light render
    - (_These visitors never have preferences set, so I think they're effectively the same as repeat visitors with unset preferences, but that's only true if unset is distinct from system, i.e., if we avoid setting the cookie to anything until the user interacts, vs creating the cookie with the default value of system on the first request; **this in particular would be good to specify, when the cookie is set and whether a fallback value is ever necessary for completely unset cookies**_)
  - Repeat visitor
    - prefers-color-scheme: dark -> dark render
      - unset
      - system
    - prefers-color-scheme: light (or unset) -> light render
      - unset
      - system
    - prefers-color-scheme: dark; manually selected light preference -> light render
    - prefers-color-scheme: light; manually selected dark preference -> dark render
- CSR (i.e., SPA navigation/setting toggles)
  - prefers-color-scheme: dark -> dark render
    - unset
    - system
  - prefers-color-scheme: light -> light render
    - unset
    - system
  - prefers-color-scheme: dark; manually selected light preference -> light render
  - prefers-color-scheme: light; manually selected dark preference -> dark render

Reply:
I ultimately felt that we didn't need a chart but have tried to supply enough new documentation that the cases should be clear. If you still think this needs further clarification in the document let me know and I'll happily amend.